### ISSUE TITLE
$joinRelation forces inner join - is there a $leftJoinRelation?

### ISSUE THREAD

**Issue by pskfry** [feathersjs-ecosystem/feathers-objection#143]
hello all - first of all thank you for the work you do it is greatly appreciated.

the issue i'm having is pretty straightforward. i have a complex graph query that utilizes $joinRelation to join a few nested relations which is working great.

the only issue i'm having is that it's using an inner join under the hood and i'd very much prefer it to be using a left join. is this possible? for the record, what i'm doing cannot be accomplished with $eager alone - i need $joinRelation because i need to also filter based on the relation's value. here's what the query looks like:

```
Person {
  name: string;
  job: Job; // Model.BelongsToOneRelation
}

Job {
  title: string;
  skill: Skill; // Model.BelongsToOneRelation
}

Skill {
  name: string;
}
```

```
api.services("persons").find({
  query: {
    $eager: "[job.[skill]]"
    $joinRelation: "[job.[skill]]"
  }
});
```

So currently, if a given Person does not have a Job, or they have a Job and that Job does not have a Skill, that Person will not show up in the query results. The reason I need $joinRelation is so I can do this (which works great!):

```
api.services("persons").find({
  query: {
    $eager: "[job.[skill]]",
    $joinRelation: "[job.[skill]]",
    "job:skill.name": userInputtedSkill
  }
});
```

I tried $leftJoinRelation with no luck - I also tried `eagerOptions` in the service options but Typescript wouldn't accept the property. Any help is greatly appreciated!

**Comment by dekelev** [comment]
Instead of adding the $joinRelation, you can use $eager with $modifyEager
or $eager with modifier. The modifiers is ObjectionJS model method that can
filter the eager results by modifying the Knex query builder object before
it's applied.

Check the docs for syntax examples.

**Comment by pskfry** [comment]
well - i just spent a few hours trying to get modifiers to work the way i want them to, but it just messed with the pagination mechanism so much that i think i need to do some minor refactoring.

i suppose this is why the objection documentation warns us that graph queries are not a silver bullet!

**Comment by dekelev** [comment]
Right, pagination issue regarding modifiers is a known issue. It was
discussed here in previous GitHub issues.

**Comment by dekelev** [comment]
Checkout modifierFiltersResults

**Comment by wz5899** [comment]
I added the $leftJoinRelation and submitted a pull request #181.

**Comment by sokulski** [comment]
> well - i just spent a few hours trying to get modifiers to work the way i want them to, but it just messed with the pagination mechanism so much that i think i need to do some minor refactoring.

Did you ever end up finding a way to make this work?

@wz5899's $leftJoinRelation seems promising. I've been wracking my brain trying to get $modifyEager to work, but I can only seem to get it to modify which children appear in the relationship, not which parents.
